### PR TITLE
`#[doc(hidden)]` can sometimes make glob re-exports not be public API. (#1072)

### PR DESCRIPTION
**Pull request by obi1kenobi**
If a glob re-export is `#[doc(hidden)]` and not deprecated, items
accessible *directly* through it are not public API because removing
that glob re-export shouldn't be breaking.

However, there's a subtlety: it's possible to publicly re-export the
names made visible by a `#[doc(hidden)]` glob re-export, making those
items public API. This results in some nasty edge cases, so we're adding
lots of test cases to cover them thoroughly.

### CODE REVIEW

File: src/visibility_tracker.rs [hunk 1]
```diff
@@ -171,23 +171,109 @@ impl<'a> VisibilityTracker<'a> {
                 currently_doc_hidden,
                 currently_deprecated,
             ));
-        } else if let Some(visible_parents) = self.visible_parent_ids.get(&next_id) {
-            for parent_id in visible_parents.iter().copied() {
+        } else if let Some(visible_parents) = self.visible_parent_edges.get(&next_id) {
+            for parent_edge in visible_parents.iter().copied() {
```
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/visibility_tracker.rs:175]: Good catch! Fixed here: https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/1074
